Objects/frameobject.c

```diff
@@ -961,15 +961,39 @@ PyFrame_FastToLocalsWithError(PyFrameObject *f)
            former here and will later use the cell for the variable.
         */
         if (kind & CO_FAST_LOCAL && kind & CO_FAST_CELL) {
-            assert(fast[i] == NULL);
             continue;
         }
 
         PyObject *name = PyTuple_GET_ITEM(co->co_localsplusnames, i);
         PyObject *value = fast[i];
-        if (kind & (CO_FAST_CELL | CO_FAST_FREE) && value != NULL) {
-            assert(PyCell_Check(value));
-            value = PyCell_GET(value);
+        if (kind & (CO_FAST_CELL | CO_FAST_FREE)) {
```

Comment:
This has a bug. If a value is passed as an argument to a function for which the matching parameter escapes, then the value will still need to boxed, even though it is non NULL.

I think you two options here:
1. Walk forward through the bytecode while `code[f->f_lasti+1] == MAKE_CELL`.
2. Special case `MAKE_CELL` in `tracing_dispatch:` (in ceval.c) and don't trace even if `f_trace_opcodes` is set.

Reply:
In this PR an escaping arg still has a separate localplus index for the cell.  That's changing in the next PR (see https://github.com/python/cpython/pull/26396#discussion_r645694620).  In that case I agree with your point.  For this PR I believe everything is right, though I'll probably make an adjustment relative to one of your other review comments.  Perhaps I misunderstood your concern?

Reply:
As to need for the value be boxed, I'm not clear on why it matters.  `MAKE_CELL` guarantees that it will be before any `*_DEREF` instruction executes since it executes at the start.  The behavior in `PyFrame_FastToLocals()` (and `PyFrame_LocalsToFast()`) does the right thing either way.  So it makes sense to let `MAKE_CELL` be in charge of the boxing.

FWIW, the only time arg value won't be boxed yet is if `MAKE_CELL` hasn't run yet, which can only happen if tracing triggers (or some C code calls `PyFrame_FastToLocals()`) before `MAKE_CELL` executes.  `MAKE_CELL` is always at the very beginning of the instructions so that reduces the likelihood.

Reply:
Ah, I think I see what you mean about the `value != NULL` case.